### PR TITLE
[REVIEW] Client Server Encryption using key stored in TPM

### PR DESCRIPTION
**Pull request by opcua-tsn-team-kalycito**
- Encrypt the private key using the key stored in TPM and remove the unencrypted private key from the filesystem
 - Used the encrypted key intermittently for software-based encryption/decryption

### CODE REVIEW

File: examples/CMakeLists.txt [hunk 1]
```diff
@@ -170,6 +170,14 @@ if(UA_ENABLE_ENCRYPTION OR UA_ENABLE_ENCRYPTION STREQUAL "MBEDTLS" OR UA_ENABLE_
     add_example(client_encryption encryption/client_encryption.c)
     target_include_directories(server_encryption PRIVATE "${PROJECT_SOURCE_DIR}/examples")
     target_include_directories(client_encryption PRIVATE "${PROJECT_SOURCE_DIR}/examples")
+    if(UA_ENABLE_ENCRYPTION_TPM_KEYSTORE)
+        add_example(server_encryption_tpm_keystore encryption/server_encryption_tpm_keystore.c)
+        add_example(client_encryption_tpm_keystore encryption/client_encryption_tpm_keystore.c)
+        target_link_libraries(server_encryption_tpm_keystore tpm2_pkcs11 ssl crypto)
+        target_link_libraries(client_encryption_tpm_keystore tpm2_pkcs11 ssl crypto)
+        target_include_directories(server_encryption_tpm_keystore PRIVATE "${PROJECT_SOURCE_DIR}/examples")
```
Comment (jpfr) [inline examples/CMakeLists.txt:178]: Those include folders are not required.

File: tools/tpm_keystore/cert_encrypt_tpm.c [hunk 2]
```diff
@@ -0,0 +1,503 @@
+/* This work is licensed under a Creative Commons CCZero 1.0 Universal License.
+ * See http://creativecommons.org/publicdomain/zero/1.0/ for more information.
+ *
+ * Copyright (c) 2021 Kalycito Infotech Private Limited
+ */
+
+/* gcc cert_encrypt_tpm.c -o cert_encrypt_tpm -ltpm2_pkcs11 -lssl -lcrypto */
+
+#include <stdio.h>
```
Comment (jpfr) [inline tools/tpm_keystore/cert_encrypt_tpm.c:9]: Add this to the tools build.
And activate in the CI.

Give this a —help option that explains the command line options.

File: CMakeLists.txt [hunk 3]
```diff
@@ -294,6 +294,14 @@ endif()
 
 option(UA_FORCE_WERROR "Force compilation with -Werror" ON)
 
+option(UA_ENABLE_ENCRYPTION_TPM_KEYSTORE "Enable TPM hardware for key encryption" OFF)
```
Comment (jpfr) [inline CMakeLists.txt:297]: Describe the flag in the build documentation.

say explicitly, that this feature adds examples and tools.
But does not change the core library.

File: CMakeLists.txt [hunk 4]
```diff
@@ -294,6 +294,14 @@ endif()
...
+mark_as_advanced(UA_ENABLE_ENCRYPTION_TPM_KEYSTORE)
+if(UA_ENABLE_ENCRYPTION_TPM_KEYSTORE)
+    if(NOT UA_ENABLE_ENCRYPTION)
```
Comment (jpfr) [inline CMakeLists.txt:300]: Wouldn’t you want to use find_library to find tpm tools and the include folder?

**Comment by jpfr** [comment]
Reuse the build option and library finding from the recent TPM PR.
option(UA_ENABLE_TPM2_PKCS11 "Enable for encryption of the PubSub payload using TPM hardware" OFF)

Otherwise good.

**Comment by JanSurft** [comment]
Great work, please resolve the conflict on the CMakeLists.txt via rebase on the current master. Then we are good to go!

File: examples/encryption/client_encryption_tpm_keystore.c [hunk 5]
```diff
@@ -0,0 +1,424 @@
...
+ *
+ *    Copyright 2021 (c) Kalycito Infotech Private Limited
+ *
+ */
+
+#include <open62541/client_config_default.h>
+#include <open62541/client_highlevel.h>
+#include <open62541/plugin/log_stdout.h>
+#include <open62541/plugin/securitypolicy.h>
+#include <open62541/server.h>
```
Comment (basyskom-meerkoetter) [inline examples/encryption/client_encryption_tpm_keystore.c:12]: why are the server includes included in the client?

**Comment by jpfr** [comment]
Nice, congrats for this nice feature!

**Pull request merged**